### PR TITLE
Enable oidc auth

### PR DESCRIPTION
**Pull request by dmyerscough**

### CODE REVIEW

**Comment by mbohlool** [comment]
To pass tests, you need a PR in the main repo with only adding dependency (not updating /base submodule).

**Comment by dmyerscough** [comment]
@mbohlool thank you, I updated the PR for (https://github.com/kubernetes-incubator/client-python/pull/331) removing the submodule update.

**Comment by dmyerscough** [comment]
Re-run test

**Comment by mbohlool** [comment]
look like there is only formatting issues with test: https://travis-ci.org/kubernetes-client/python-base/jobs/270071663

Sorry that I didn't spend time on this yet because of the rush of PRs before kubernetes code freeze. I will do a proper review very soon.

**Comment by dmyerscough** [comment]
Thanks @mbohlool Ill write some unit tests this week or next (time depending).

**Comment by mbohlool** [comment]
any update on this? also did you test this? e.g. is there a `minikube` way to setup oidc and test this? or what flavor of kubernetes cluster have oidc support?

**Comment by dmyerscough** [comment]
@mbohlool sorry I have been down with the cold.. I have tested this with CoreOS Tectonic which uses OIDC for authentication.

I can probably just write unit tests to begin with before integration tests.

**Comment by mbohlool** [comment]
@dmyerscough no worries. Integration tests are not required for this PR to be merged, but if you are up to add it, go for it. But unit tests are more important.

Thanks for doing this.

File: config/kube_config_test.py [hunk 1]
```diff
@@ -58,6 +60,63 @@ def _raise_exception(st):
 # token for me:pass
 TEST_BASIC_TOKEN = "Basic bWU6cGFzcw=="
 
+TEST_OIDC_LOGIN = (
```
Comment (mbohlool) [inline config/kube_config_test.py]: I don't think we should use the actual login keys in test. Can you mock it and use something like "test_login..."

Comment (dmyerscough) [reply config/kube_config_test.py]: I can mock this out; I noticed due to the mocking the code coverage test fails.

**Comment by mbohlool** [comment]
@dmyerscough friendly ping. Any updates?

**Comment by mbohlool** [comment]
@dmyerscough you may want to look at this comment https://github.com/kubernetes-incubator/client-python/issues/368#issuecomment-337778192

**Comment by dmyerscough** [comment]
@mbohlool I have been pretty busy and haven't had time to look at this.

File: config/kube_config.py [hunk 2]
```diff
@@ -217,6 +225,93 @@ def _load_user_token(self):
...
+                'client_secret': provider['config']['client-secret']
+            },
+            auto_refresh_url=response['token_endpoint']
+        )
+
+        try:
+            refresh = request.refresh_token(
+                token_url=response['token_endpoint'],
+                refresh_token=provider['config']['refresh-token'],
+                verify=ca_cert.name
```
Comment (lichen2013) [inline config/kube_config.py]: Using `auth` here can remove all the use of cert staff above:

```
refresh = request.refresh_token(
    token_url=response['token_endpoint'],
    refresh_token=provider['config']['refresh-token'],
    auth=(provider['config']['client-id'], provider['config']['client-secret']))
```

File: config/kube_config.py [hunk 3]
```diff
@@ -217,6 +225,93 @@ def _load_user_token(self):
...
+        if len(parts) != 3:  # Not a valid JWT
+            return None
+
+        if PY3:
+            jwt_attributes = json.loads(
+                base64.b64decode(parts[1]).decode('utf-8')
+            )
+        else:
+            jwt_attributes = json.loads(
+                base64.b64decode(parts[1])
```
Comment (lichen2013) [inline config/kube_config.py]: I get an error in my test:

> TypeError: Incorrect padding

Would be nice to handle this error here.

**Comment by ltamaster** [comment]
Hi @dmyerscough , @mbohlool:
I am interested in this enhancement, I forked the last version of the client and apply manually these changes. 
I added a fix for `TypeError: Incorrect padding` and I changed the variable values on the test.
Please let me know if you are OK if I send a new PR.

Thanks
Luis

**Comment by dmyerscough** [comment]
@ltamaster you are more than welcome to take over this, I have been swamped with a new job and preparing for a new born.

**Comment by mbohlool** [comment]
@ltamaster I am fine with a new PR.

**Comment by ltamaster** [comment]
Hi @dmyerscough,
I think I am ready for the PR, my only problem is that I cannot pass the SSL certificate to the refresh_token  call. According to this PR (https://github.com/requests/requests-oauthlib/pull/293) it is possible, but it didn't work for me. My workaround is not to verify the certificate using `request.verify=false` ( on the call when the client try to refresh the token). Is that OK  @mbohlool @dmyerscough?

**Comment by ltamaster** [comment]
@dmyerscough  @mbohlool  forgot my last message, I found my problem

**Comment by ltamaster** [comment]
hi @dmyerscough 
I have problems with travis on the test `test_oidc_with_refresh` (on coverage), any idea?

```ERROR: test_oidc_with_refresh (kubernetes.config.kube_config_test.TestKubeConfigLoader)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/tmp.yxT81y67iD/python/.tox/coverage/lib/python2.7/site-packages/mock/mock.py", line 1297, in patched
    arg = patching.__enter__()
  File "/tmp/tmp.yxT81y67iD/python/.tox/coverage/lib/python2.7/site-packages/mock/mock.py", line 1353, in __enter__
    self.target = self.getter()
  File "/tmp/tmp.yxT81y67iD/python/.tox/coverage/lib/python2.7/site-packages/mock/mock.py", line 1523, in <lambda>
    getter = lambda: _importer(target)
  File "/tmp/tmp.yxT81y67iD/python/.tox/coverage/lib/python2.7/site-packages/mock/mock.py", line 1206, in _importer
    thing = __import__(import_path)
ImportError: No module named config

```

**Comment by mvle** [comment]
I apologize in advance if this is already supported, but from looking at the PR, I think there needs to be a way to refresh the token not only at configuration initialization time but also during invocation of the K8 API calls. The token might expire right before a call thus causing the call to fail with an authentication failure. This will be needed for services that run a long time and rely on being able to reach the K8 Api Server (also short services that happen to be unlucky).

I think this can be done with a check to 'exp' field in the id-token right before a call_api or catch a 401 return status and perform a refresh token operation and update the configuration object with the new id_token and refresh_token values from the OAuth server.

**Comment by dmyerscough** [comment]
Please ref #48

**Pull request closed without merging**